akka-http-core/src/main/scala/akka/http/scaladsl/model/Uri.scala

```diff
@@ -909,14 +914,28 @@ object UriRendering {
   }
 
   private[http] def encode(r: Rendering, string: String, charset: Charset, keep: CharPredicate,
-                           replaceSpaces: Boolean = false): r.type = {
+                           replaceSpaces:             Boolean = false,
+                           detectAlreadyEscapedChars: Boolean = false): r.type = {
     val asciiCompatible = isAsciiCompatible(charset)
     @tailrec def rec(ix: Int): r.type = {
-      def appendEncoded(byte: Byte): Unit = r ~~ '%' ~~ CharUtils.upperHexDigit(byte >>> 4) ~~ CharUtils.upperHexDigit(byte)
+      def appendEncoded(byte: Byte): Unit =
+        r ~~ '%' ~~ CharUtils.upperHexDigit(byte >>> 4) ~~ CharUtils.upperHexDigit(byte)
+
       if (ix < string.length) {
         val charSize = string.charAt(ix) match {
-          case c if keep(c)                     ⇒ { r ~~ c; 1 }
-          case ' ' if replaceSpaces             ⇒ { r ~~ '+'; 1 }
+          case c if keep(c)         ⇒ { r ~~ c; 1 }
+          case ' ' if replaceSpaces ⇒ { r ~~ '+'; 1 }
+          case '%' if detectAlreadyEscapedChars ⇒ {
+            var c1: Char = 0
+            var c2: Char = 0
+            val isEscapedChar = string.length > ix + 2 && {
+              c1 = string.charAt(ix + 1)
+              c2 = string.charAt(ix + 2)
+              c1 < 127 && c2 < 127
+            }
+            if (isEscapedChar) { r ~~ '%' ~~ c1 ~~ c2; 3 }
+            else { appendEncoded('%'.toByte); 1 }
+          }
```

Comment:
I'm very torn about this actually... This new logic is used only in the authority really.
The reason the authority is assumed to be escaped already is that otherwise we'd get double escapes